### PR TITLE
Implemented issue #870

### PR DESCRIPTION
**Pull request by Soarc**
Fixes # .

### Checklist

- [*] I have included examples or tests
- [*] I have updated the change log
- [*] I am listed in the CONTRIBUTORS file
- [*] I have cleaned up the commit history (use rebase and squash)

Implemented issue #870 .
Added CropGridlines property to Axis that indicated, if Gridlines should be cropped by perpendicular axes.

### CODE REVIEW

File: Source/OxyPlot.SharpDX.WPF45/PlotImage.cs [hunk 1]
```diff
@@ -329,7 +329,15 @@ void InitRendering()
 
        
 
-            _d3d11Device = new D3D11Device(DriverType.Hardware, DeviceCreationFlags.BgraSupport, FeatureLevel.Level_11_0, FeatureLevel.Level_10_1);
+            _d3d11Device = new D3D11Device(
```
Comment (objorke) [inline Source/OxyPlot.SharpDX.WPF45/PlotImage.cs]: this change should be in a separate PR. please revert

**Comment by Soarc** [comment]
Reverted

File: Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs [hunk 2]
```diff
@@ -322,13 +332,35 @@ protected virtual void RenderMajorItems(Axis axis, double axisPosition, double t
                 {
                     if (isHorizontal)
                     {
-                        majorSegments.Add(new ScreenPoint(transformedValue, plotAreaTop));
-                        majorSegments.Add(new ScreenPoint(transformedValue, plotAreaBottom));
+                        if (!cropGridlines)
```
Comment (objorke) [inline Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs]: it looks like this code block is repeated 6 times. Could it be a good candidate for a method or anonymous method?

File: Source/OxyPlot.SharpDX.WPF45/PlotImage.cs [hunk 3]
```diff
@@ -590,4 +590,4 @@ public Rect MakeVisible(Visual visual, Rect rectangle)
             //TODO: implement this????????
         }
     }
-}
+}
```
Comment (objorke) [inline Source/OxyPlot.SharpDX.WPF45/PlotImage.cs]: not properly reverted

File: Source/Examples/ExampleLibrary/Axes/AxisExamples.cs [hunk 4]
```diff
@@ -1151,5 +1151,59 @@ private static PlotModel CreateTickStyleExample(TickStyle tickStyle)
...
+            plotModel1.Axes.Add(new LinearAxis
+            {
+                MajorGridlineStyle = LineStyle.Solid,
+                MinorGridlineStyle = LineStyle.Dot,
+                ExtraGridlines = new double[] { 46d },
+                ExtraGridlineColor = OxyColors.Red,
+                StartPosition = 0.1,
+                EndPosition = 0.4,
+                CropGridlines = true
+
```
Comment (objorke) [inline Source/Examples/ExampleLibrary/Axes/AxisExamples.cs]: blank lines

File: Source/Examples/ExampleLibrary/Axes/AxisExamples.cs [hunk 5]
```diff
@@ -1151,5 +1151,59 @@ private static PlotModel CreateTickStyleExample(TickStyle tickStyle)
...
+            plotModel1.Axes.Add(new LinearAxis
+            {
+                MajorGridlineStyle = LineStyle.Solid,
+                MinorGridlineStyle = LineStyle.Dot,
+                ExtraGridlines = new double[] { 46d },
+                ExtraGridlineColor = OxyColors.Red,
+                StartPosition = 0.6,
+                EndPosition = 0.9,
+                CropGridlines = true
+
```
Comment (objorke) [inline Source/Examples/ExampleLibrary/Axes/AxisExamples.cs]: blank line

File: Source/Examples/ExampleLibrary/Axes/AxisExamples.cs [hunk 6]
```diff
@@ -1151,5 +1151,59 @@ private static PlotModel CreateTickStyleExample(TickStyle tickStyle)
...
+            {
+                MajorGridlineStyle = LineStyle.Solid,
+                MinorGridlineStyle = LineStyle.Dot,
+                Position = AxisPosition.Bottom,
+                ExtraGridlines = new double[] { 46d },
+                ExtraGridlineColor = OxyColors.Red,
+                StartPosition = 0.6,
+                EndPosition = 0.9,
+                CropGridlines = true
+
```
Comment (objorke) [inline Source/Examples/ExampleLibrary/Axes/AxisExamples.cs]: blank line

File: Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs [hunk 7]
```diff
@@ -294,6 +295,15 @@ protected virtual void RenderMajorItems(Axis axis, double axisPosition, double t
             var perpendicularAxis = axis.IsHorizontal() ? this.Plot.DefaultYAxis : this.Plot.DefaultXAxis;
             var dontRenderZero = axis.PositionAtZeroCrossing && perpendicularAxis.PositionAtZeroCrossing;
 
+            List<Axis> perpAxes = null;
+            if (cropGridlines)
+            {
+                if (isHorizontal)
+                    perpAxes = this.Plot.Axes.Where(x => x.IsXyAxis() && x.IsVertical()).ToList();
```
Comment (objorke) [inline Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs]: this line should be enclosed in {} - stylecop rule

File: Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs [hunk 8]
```diff
@@ -294,6 +295,15 @@ protected virtual void RenderMajorItems(Axis axis, double axisPosition, double t
...
+                else
+                    perpAxes = this.Plot.Axes.Where(x => x.IsXyAxis() && x.IsHorizontal()).ToList();
```
Comment (objorke) [inline Source/OxyPlot/Axes/Rendering/HorizontalAndVerticalAxisRenderer.cs]: enclose in {}

File: Source/Examples/ExampleLibrary/Axes/AxisExamples.cs [hunk 9]
```diff
@@ -1151,5 +1151,59 @@ private static PlotModel CreateTickStyleExample(TickStyle tickStyle)
             plotModel1.Axes.Add(new LinearAxis { Position = AxisPosition.Bottom, TickStyle = tickStyle });
             return plotModel1;
         }
+        
+        [Example("Gridlines Cropping: Horizontal and vertical")]
```
Comment (objorke) [inline Source/Examples/ExampleLibrary/Axes/AxisExamples.cs]: Good!

**Comment by Soarc** [comment]
Thanks for comments. I'll fix this soon. 
P.S. If i squash this branch, how github will perform?

**Comment by objorke** [comment]
squash + `git push -f` should work fine.

**Comment by Soarc** [comment]
I've implemented changes, that you suggested. But can't do proper squash (I really don't understand, how this should be done).
I can close this pull request and create new one with clean history. Should i do this?

**Comment by objorke** [comment]
Thanks! I can use GitHub to squash before merging the PR.
